### PR TITLE
feat(rust): Adds the ability to specify custom derive attributes

### LINKED ISSUE #554: Add support for Serialize, Deserialize, and Default
The structures that we get from Rust when doing wit-bindgen do not have the `Default` or `Serialize` or `Deserialize` derivations on them. This means that as soon as I attempt to marshal one of these data types somewhere without using a wit function call, I have to manually create my own "mirror types" that have the same structure, but with the serialization attributes on them. Then I have to manually implement either `From` to convert the wit-bindgen type to my type prior to serialization. 

I understand that this kind of activity typically shows up in anti-corruption layers and in an ideal world, the "outside" model and the inside (wit-bindgen) model would never overlap. However, if I have a _lot_ of types and I need to marshal those in and out of my components without making bindgen function calls, I'm stuck redefining every type in my wit world.

Can we get this functionality? If it isn't something that you want on by default, it could be added as a flag to the macro like this:

```rust
wit_bindgen::generate!({
    path: "../wit",
    world: "bankaccount-aggregate",
    serializers: true
});
```

Support for the `serializers` flag could even be hidden behind a feature flag on the wit-bindgen crate.

### PR DESCRIPTION
**Pull request by thomastaylor312**
I also did a little bit of cleanup of clippy warnings while I was in these files.

Closes #554

### CODE REVIEW

**Comment by thomastaylor312** [comment]
I am pretty sure the test failure is a flake or not related to my code as I am seeing it in other PRs as well

File: crates/rust-lib/src/lib.rs [hunk 1]
```diff
@@ -159,7 +160,7 @@ pub trait RustGenerator<'a> {
         param_mode: TypeMode,
         sig: &FnSig,
     ) -> Vec<String> {
-        let params = self.print_docs_and_params(func, param_mode, &sig);
+        let params = self.print_docs_and_params(func, param_mode, sig);
```
Comment (alexcrichton) [inline crates/rust-lib/src/lib.rs:163]: No worries for this PR, but for future contributions I'd recommend splitting out clippy changes into a separate commit or possibly PR. Makes the "meat" elsewhere a bit easier to review that way.

File: crates/rust-lib/src/lib.rs [hunk 2]
```diff
@@ -572,8 +573,11 @@ pub trait RustGenerator<'a> {
         record: &Record,
         docs: &Docs,
         derive_component: bool,
+        additional_derives: Vec<String>,
```
Comment (alexcrichton) [inline crates/rust-lib/src/lib.rs:576]: It's a bit unfortunate to pass this around in so many places. In theory this `rust-lib` crate shouldn't even exist any longer. This was only split out as it was shared between Wasmtime and a Rust guest, but Wasmtime no longer lives here nor uses this so it's an unnecessary abstraction nowadays. Additionally that's a large enough refactor there's no need to do it in this PR.

In the meantime though a possible alternative design would be a trait method on `RustGenerator` along the lines of `fn additional_derives(&self) -> &[String]`. I think that should avoid most of the passing around and clean that up slightly. Mind taking this on and I'll as a follow up work on merging this crate into the `rust` crate?

File: crates/rust-lib/src/lib.rs [hunk 3]
```diff
@@ -586,12 +590,20 @@ pub trait RustGenerator<'a> {
...
+                derives.extend(["Copy", "Clone"].into_iter().map(|s| s.to_string()));
             } else if info.is_clone() {
-                self.push_str("#[derive(Clone)]\n");
+                derives.insert("Clone".to_string());
+            }
+            if !derives.is_empty() {
+                self.push_str("#[derive(");
+                // NOTE(thomastaylor312): Since this is code generation, I don't mind the iteration
+                // and allocation of a new vec here. If it ever matters, we can create a function
+                // that glues it together via iterator
```
Comment (alexcrichton) [inline crates/rust-lib/src/lib.rs:604]: No worries on leaving a note here, very little of wit-bindgen is optimized for performance :)

File: crates/rust-lib/src/lib.rs [hunk 4]
```diff
@@ -572,8 +573,11 @@ pub trait RustGenerator<'a> {
         record: &Record,
         docs: &Docs,
         derive_component: bool,
+        additional_derives: Vec<String>,
     ) {
         let info = self.info(id);
+        // We use a hash set to make sure we don't have any duplicates
+        let additional_derives: HashSet<String> = additional_derives.into_iter().collect();
```
Comment (alexcrichton) [inline crates/rust-lib/src/lib.rs:580]: One important caveat to be aware of about `HashSet<T>` is that its iteration order is not defined and not deterministic between runs of a process. This means, for example, that different code is generated from the same WIT source as derives might be shuffled around.

Mind using a `BTreeSet` (or `IndexSet`) here instead to preserve order?

File: crates/rust-lib/src/lib.rs [hunk 5]
```diff
@@ -853,6 +878,7 @@ pub trait RustGenerator<'a> {
         }
     }
 
+    #[allow(clippy::too_many_arguments)]
```
Comment (alexcrichton) [inline crates/rust-lib/src/lib.rs:881]: This project doesn't use clippy, so mind leaving this out?

**Review by alexcrichton — APPROVED** [review]
Thanks for this! No worries on the failed test, that's [known to be flaky](https://github.com/bytecodealliance/wit-bindgen/pull/680) and should be resolved on `main` soon through ignoring it.

At a high level one concern I have about this is that it won't work for `Resource` types. That won't implement Serialize/Deserialize and doesn't currently implement PartialEq/Hash, although it perhaps could. This means that implementing traits uniformly for all types may not be sufficient and eventually this may need a form of "ignore this type" or "only derive for this type". That all being said I think it's fine to deal with those situations as they come up later, for now this is suitable and works well so I'd say we should land it and then iterate on further syntaxes later if necessary.

File: crates/rust-lib/src/lib.rs (hunk 2 above)
Comment (thomastaylor312) [reply crates/rust-lib/src/lib.rs:576]: Yep, I'll shift it around. Honestly was having a hard time keeping track of where everything was at during the beginning. Now that I have a better grasp, I can do this quick

File: crates/rust-lib/src/lib.rs (hunk 4 above)
Comment (thomastaylor312) [reply crates/rust-lib/src/lib.rs:580]: Oh right, I should have though about the ordering issue. I'll switch that around

File: crates/rust-lib/src/lib.rs (hunk 5 above)
Comment (thomastaylor312) [reply crates/rust-lib/src/lib.rs:881]: Yeah I can leave it out

**Comment by thomastaylor312** [comment]
@alexcrichton Should be ready to go for a rereview

In regards to this:

> At a high level one concern I have about this is that it won't work for Resource types. That won't implement Serialize/Deserialize and doesn't currently implement PartialEq/Hash, although it perhaps could. This means that implementing traits uniformly for all types may not be sufficient and eventually this may need a form of "ignore this type" or "only derive for this type". That all being said I think it's fine to deal with those situations as they come up later, for now this is suitable and works well so I'd say we should land it and then iterate on further syntaxes later if necessary.

I completely agree with the path forward here. I was curious how this would interact with resources, but it definitely seemed easiest to add this to struct/enum types for now and then we can see if there is a way to make it more specific in the future

File: crates/rust-lib/src/lib.rs (hunk 4 above)
Comment (thomastaylor312) [reply crates/rust-lib/src/lib.rs]: This should be all resolved, but leaving the comment unresolved until you have a chance to take a look at it

**Comment by alexcrichton** [comment]
Looks great to me, thanks!

One option is to use the `TypeInfo` bits to avoid these extra derives for types-with-resources, but that's still a bit specific and may not work out in all cases. In any case this looks good for now 👍

**Comment by sehz** [comment]
does this work work in `cargo component` ?

**Comment by thomastaylor312** [comment]
@sehz I was actually going to go open an issue over there in cargo component

**Pull request merged**